Composer/packages/server/src/externalContentProvider/powerVirtualAgentsProvider.ts

```diff
@@ -0,0 +1,173 @@
+
+const log = logger.extend('pva-provider');
+
+const COMPOSER_1P_APP_ID = 'ce48853e-0605-4f77-8746-d70ac63cc6bc';
+
+export type PowerVirtualAgentsMetadata = ContentProviderMetadata & {
+  baseUrl: string;
+  botId: string;
+  dialogId?: string;
+  envId: string;
+  name: string;
+  tenantId: string;
+  triggerId?: string;
+};
+
+// TODO: these need to be set to prod
+const authCredentials = {
+  // web auth flow
+  clientId: COMPOSER_1P_APP_ID,
+  scopes: ['a522f059-bb65-47c0-8934-7db6e5286414/.default'], // int / ppe
+  // electron auth flow
+  targetResource: 'a522f059-bb65-47c0-8934-7db6e5286414',
+};
+
+const getBaseUrl = () => {
+  const pvaEnv = (process.env.COMPOSER_PVA_ENV || '').toLowerCase();
+  switch (pvaEnv) {
+    case 'prod': {
+      const url = 'https://powerva.microsoft.com/api/botmanagement/v1';
+      log('PROD env detected, grabbing PVA content from %s', url);
+      return url;
+    }
+
+    case 'ppe': {
+      const url = 'https://bots.ppe.customercareintelligence.net/api/botmanagement/v1';
+      log('PPE env detected, grabbing PVA content from %s', url);
+      return url;
+    }
+
+    case 'int': {
+      const url = 'https://bots.int.customercareintelligence.net/api/botmanagement/v1';
+      log('INT env detected, grabbing PVA content from %s', url);
+      return url;
+    }
+
+    default: {
+      const url = 'https://bots.int.customercareintelligence.net/api/botmanagement/v1';
+      log('No env flag detected, grabbing PVA content from %s', url);
+      return url;
+    }
+  }
+};
+
+function prettyPrintError(err: string | any): string {
+  if (typeof err === 'string') {
+    return err;
+  }
+  return JSON.stringify(err, null, 2);
+}
+
+export class PowerVirtualAgentsProvider extends ExternalContentProvider<PowerVirtualAgentsMetadata> {
+  private tempBotAssetsDir = join(process.env.COMPOSER_TEMP_DIR as string, 'pva-assets');
+
+  constructor(metadata: PowerVirtualAgentsMetadata) {
+    super(metadata);
+  }
+
+  public async downloadBotContent(): Promise<BotContentInfo> {
+    try {
+      // fetch zip
+      const url = this.getContentUrl();
+      const options: RequestInit = {
+        method: 'GET',
+        headers: await this.getRequestHeaders(),
+      };
+      const result = await fetch(url, options);
+
+      const eTag = result.headers.get('etag') || '';
+      const contentType = result.headers.get('content-type');
+      if (!contentType || contentType !== 'application/zip') {
+        const json = await result.json();
+        throw `Did not receive zip back from PVA: ${prettyPrintError(json)}`;
+      }
+
+      // write the zip to disk
+      if (result && result.body) {
+        ensureDirSync(this.tempBotAssetsDir);
+        const zipPath = join(this.tempBotAssetsDir, 'bot-assets.zip');
+        const writeStream = createWriteStream(zipPath);
+        await new Promise((resolve, reject) => {
```

Comment:
Is it fine to overwrite an existing one?

Reply:
Yes, but I'll timestamp it just to be careful